server/services/zwave/lib/comClass/defaultCommandClass.js

```diff
@@ -0,0 +1,89 @@
+const { transformValue, normalizeValue } = require('../utils/valueBinders');
+
+/**
+ * @description Get the command class ID.
+ *
+ * @returns {number} - Command Class ID.
+ *
+ * @example commandClass.getId();
+ */
+function getId() {
+  return -1;
+}
+
+/**
+ * @description Get the value following a value changed event.
+ *
+ * @param {Object} node - The node the value changed occurs.
+ * @param {Object} valueChanged - The value changed.
+ *
+ * @returns {Object} - A zWave value.
+ *
+ * @example
+ *     const changedValue = comClass.getChangedValue(node, value);
+ */
+function getChangedValue(node, valueChanged) {
+  return valueChanged;
+}
+
+/**
+ * @description Transform a value according to the Command Class.
+ *
+ * @param {Object} node - Node of the value.
+ * @param {number} comClass - Command Class Id of the value.
+ * @param {number} index - Index of the value.
+ * @param {number} instance - Instance of the value.
+ * @param {any} value - Value to transform.
+ *
+ * @returns {any} The transformed value.
+ *
+ * @example comClass.getTransformedValue(node, 110, 0, 1, 50);
+ */
+function getTransformedValue(node, comClass, index, instance, value) {
+  return transformValue(node.classes[comClass][index][instance].type, value);
+}
+
+/**
+ * @description Normalize a value according to the Command Class.
+ *
+ * @param {Object} node - Node of the value.
+ * @param {number} comClass - Command Class Id of the value.
+ * @param {number} index - Index of the value.
+ * @param {number} instance - Instance of the value.
+ * @param {any} value - Value to transform.
+ *
+ * @returns {any} The normalized value.
+ *
+ * @example comClass.getNormalizedValue(node, 110, 0, 1, 50);
+ */
+function getNormalizedValue(node, comClass, index, instance, value) {
+  return normalizeValue(node.classes[comClass][index][instance].type, value);
```

Comment:
Can I ask you a question, are you a Java developer? 😄 

I begin to see how you think: everything needs to be "class" (no offense), and it leads to lots of functions like this one (with no real code inside). 

I know that you've maybe worked like that for years so don't get offended, but I feel that the code is hard to read as  the user needs to jump from file to file, and most code is useless like this one.

All this could have been a one-line of not duplicate code in one helper file, no ?

I know that you already said that "the code will be complex next", but in my experience working 7 years on Gladys, other integrations are completely fine, and coded in a functional way.

Reply:
ahah, no worries, no offense ;)

No, I'm not a Java developper :) I'm using multiple languages (recently mainly on JS, C#, PHP), but I didn't play with Java for a long time.

> All this could have been a one-line of not duplicate code in one helper file, no ?

You could always do everything in one file. It's just a POV about separation and responsabilities. Is it the responsability to a CommandClass to know how to normalize a value (something linked to Gladys itself)? Probably not. The CommandClass needs to normalize a value, but doesn't need to know how. Putting this dedicated code in a dedicated file makes it easier to test (tests are focused on this class). Intention of the class/file is clearly defined. Yes, I can put this logic in the DefaulCommandClass, but it will make it bigger and knowing some concepts not really related to a CommandClass (so a file harder to read, harder to test as it should tests some things not directly related to CommandClass and so on). 

> the code is hard to read as the user needs to jump from file to file

It depends what you are looking for. If you need to deep dive in what's going on, yes you need one more file to open. But if you're clearly focused on CommandClass behavior, you don't need to deep dive in this part of code. It's out of the CommandClass concept and you can so assume this external file makes its job as it should. It's like the `emit` function. I don't need to know how it's done. i'm using it, assuming it's doing its job as expected so I can move forward. It makes it easier to read.

Finally, it might be more a habit issue than a real one. For example, when I firstly came on Gladys, as a regular JS developper, I've been very surprised about how the backend is structured: I think separating all functions of a same class in dedicated files doesn't help at all developpers (neither IDE). It's really hard and uncommon to navigate from functions to functions this way (Imagine moving on ES6+ syntax with real classes: you would not explode your functions in different files like it's actually done). But, finally, even like that, the product is working pretty well.

If your concern is this `valueBinders` file, I can bring it back to one file... I will have some troubles to sleep correctly, but it should be fine with time ;)

Reply:
I'm just concerned about consistency around the project, we don't work like that on other integrations :) 

This debate is hard as we are both opinionated in two opposite directions: you're mostly pro-class/inheritance, and I'm mostly pro-function and against inheritance.

It's an eternal debate in programming, and there are many articles on the pro/cons of each paradigm. Any of us won't change opinion I guess aha 😄 

To give an example of others integration that did a functional approach that worked pretty well: The Philips Hue integration or the Zigbee2mqtt integration

Both have a "models" folder (See [Philips Hue models](https://github.com/GladysAssistant/Gladys/tree/master/server/services/philips-hue/lib/models) or [zigbee2mqtt models](https://github.com/GladysAssistant/Gladys/tree/master/server/services/zigbee2mqtt/model) ) . 

Each file is a function that returns a formatted device. When needed, we use helper functions or a set of constants to get min/max, that kind of stuff.

It's just my opinion and I don't want to convert you, but I find this kind of implementation easier to read, code & maintain. 

When using inheritance, class are tightly coupled together and refactoring is harder when adding a new device. One new developer needs to know all the classes before adding a new device, it's prone to more refactoring errors. 

> For example, when I firstly came on Gladys, as a regular JS developper, I've been very surprised about how the backend is structured: I think separating all functions of a same class in dedicated files doesn't help at all developpers (neither IDE).

This one is another debate, and your point is definitely valid: it's harder to autocomplete with the "one file per function" approach. But on the other hand, it's cleaner IMO to split the functions in multiple files, make it's easier to access one function by typing cmd + p and the name of the function (ex: cmd + P + `device.setValue` ) and directly accessing the function (which you can't if you have one giant files with all functions). Believe me it's not a simple decision I took, I thought about it a lot when implementing this :p